### PR TITLE
.Net: StepwisePlanner Chat support

### LINKED ISSUE #2367: Update StepwisePlanner to use ChatHistory for generating completions instead of TextCompletion with Chat models.

### LINKED ISSUE #2466: .Net: Stepwise Planner: Planner fails to mark final answer as [Final Answer]
**Describe the bug**
With some goals, the stepwise planner will eventually have a thought that should be the final answer but fails to mark it as such. It continues to repeat a thought that should be the final answer until it runs out of iterations. My specific scenario is analyzing code from our repository based on a user chat. I have noticed that this only occurs for larger code files. Smaller code files usually end up having a final answer.

Example scratch pad from the planner showing the repetition

> This was my previous work (but they haven't seen any of it! They only see what I return as final answer):
> [THOUGHT] To determine if AminoStartup configures Swagger, I need to search for relevant code references in the MTech Systems' agricultural software. I will use the CodeSearchSkill.SearchCodeByReference function to search for "AminoStartup" and "Swagger" in the codebase.
> [ACTION] {"action": "CodeSearchSkill.SearchCodeByReference","action_variables": {"input":"","searchText":"AminoStartup Swagger"}}
> [OBSERVATION] [
> 	"**file path removed**"
> ]
> [THOUGHT] I found a file named "AminoStartup.cs" that might contain the configuration for Swagger. I will now use the CodeSearchSkill.GetFileContent function to get the content of this file and check if it configures Swagger.
> [ACTION] {"action": "CodeSearchSkill.GetFileContent","action_variables": {"input":"","filePath":"**file path removed**"}}
> [OBSERVATION] 
> [FILE CONTENT]
> **FileContent would be here**
> [END FILE CONTENT]
> [THOUGHT] Yes, AminoStartup configures Swagger in the MTech Systems' agricultural software. I found a file named "AminoStartup.cs" in the codebase, and upon examining its content, I found that it includes the necessary configuration for Swagger.
> [THOUGHT] Yes, AminoStartup configures Swagger in the MTech Systems' agricultural software. I found a file named "AminoStartup.cs" in the codebase, and upon examining its content, I found that it includes the necessary configuration for Swagger.

**To Reproduce**
1. Download repro project.
2. Enter your endpoint and key in the Program.cs file.
3. Open CodeSearchSkill and paste the contents of a large code file in the variable FileContentToReturn
4. Run application and observe the stepwise planner scratch pad

**Expected behavior**
The first complete answer returned about the file should be marked as final answer

**Platform**
 - OS: Windows
 - IDE: Visual Studio
 - Language: C#
 - Source: NuGet 0.19.230804.2-preview

[ConsoleApp2.zip](https://github.com/microsoft/semantic-kernel/files/12346056/ConsoleApp2.zip)

### LINKED ISSUE #2553: StepwisePlanner -- Action regex doesn't support actions that have braces in the string values

### PR DESCRIPTION
**Pull request by lemillermicrosoft**
### Motivation and Context
This pull request introduces a series of improvements and updates to the StepwisePlanner, StepwisePlannerTests, and WebSearchEngineSkill. The changes include   
better handling of action invocation, function retrieval, logging, chat history management, execution helpers, and the addition of an offset parameter to the   
WebSearchEngineSkill. Additionally, the StepwisePlannerTests have been enhanced with more test cases and refined expected minimum steps, while the
ChatRequestSettings class has been improved with a method to create a new settings object from another settings object.

Resolves #2367
Fixes #2466 
Fixes #2553 

### Description
1. Improved the StepwisePlanner class with better action invocation, function retrieval, logging, chat history management, and execution helpers.
2. Enhanced the StepwisePlannerTests with more test cases and refined expected minimum steps.
3. Added an offset parameter to the WebSearchEngineSkill class for improved search functionality.
4. Improved the ChatRequestSettings class with a method to create a new settings object from another settings object.
5. Updated the Example51_StepwisePlanner.cs file with new questions, improved result handling, and additional output for better analysis.
6. Enhanced the ParseResultTests with new test cases for the ParseResult class.
7. Updated the WebSearchEngineSkill to include parameter descriptions for the SearchAsync method.
8. Added project references and minor changes to the SemanticKernel.MetaPackage project.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

Co-authored-by: Andrew Hesky @andhesky

### CODE REVIEW

**Comment by lemillermicrosoft** [comment]
Related to #2444

File: dotnet/src/Extensions/Planning.StepwisePlanner/Skills/StepwiseStep/config.json [hunk 1]
```diff
@@ -12,20 +12,15 @@
...
-      },
       {
         "name": "functionDescriptions",
         "description": "The manual of the agent's functions",
         "defaultValue": ""
+      },
+      {
+        "name": "suffix",
+        "description": "",
+        "defaultValue": "Let's break down the problem step by step and think about the best approach. Label steps as they are taken.\n\nContinue the thought process!"
```
Comment (teresaqhoang) [inline dotnet/src/Extensions/Planning.StepwisePlanner/Skills/StepwiseStep/config.json:23]: This is important to include even if the user wants to provide supplemental usage instructions. Otherwise, it'd be overridden and there'd be no pointers to go step by step

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 2]
```diff
@@ -33,39 +39,43 @@ public class StepwisePlanner : IStepwisePlanner
...
             string promptConfigString = EmbeddedResource.Read("Skills.StepwiseStep.config.json");
             if (!string.IsNullOrEmpty(promptConfigString))
             {
-                promptConfig = PromptTemplateConfig.FromJson(promptConfigString);
+                this._promptConfig = PromptTemplateConfig.FromJson(promptConfigString);
             }
         }
+        else
+        {
+            this._promptConfig = this.Config.PromptUserConfig;
```
Comment (teresaqhoang) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:69]: this assignment isn't needed since it's already assigned in line 54, or line 54 isn't needed

File: dotnet/samples/KernelSyntaxExamples/Example51_StepwisePlanner.cs [hunk 3]
```diff
@@ -19,30 +22,140 @@
...
+
     public static async Task RunAsync()
     {
         string[] questions = new string[]
         {
-            "Who is the current president of the United States? What is his current age divided by 2",
-            // "Who is Leo DiCaprio's girlfriend? What is her current age raised to the (his current age)/100 power?",
-            // "What is the capital of France? Who is that city's current mayor? What percentage of their life has been in the 21st century as of today?",
-            // "What is the current day of the calendar year? Using that as an angle in degrees, what is the area of a unit circle with that angle?"
+            "What is the tallest mountain on Earth? How tall is it divided by 2?",
```
Comment (lemillermicrosoft) [inline dotnet/samples/KernelSyntaxExamples/Example51_StepwisePlanner.cs:39]: Add simple examples "What color is the sky?" "What is the weather in Seattle?"

File: dotnet/src/Extensions/Planning.StepwisePlanner/Skills/StepwiseStep/config.json (hunk 1 above)
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/Skills/StepwiseStep/config.json:23]: Yeah, in the example usage I'm building on this value in the configuration. Trying to find the balance between customization and guardrails.

Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/Skills/StepwiseStep/config.json:23]: e.g. `plannerConfig.Suffix = $"{Suffix}\n{plannerConfig.Suffix}";`

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlannerConfig.cs [hunk 4]
```diff
@@ -53,10 +54,33 @@ public sealed class StepwisePlannerConfig
...
+    /// </summary>
+    public string Suffix { get; set; } = @"Let's break down the problem step by step and think about the best approach. Label steps as they are taken.
+
+Continue the thought process!";
+    #endregion Execution configuration
+
+    /// <summary>
+    /// A logger factory to use for logging.
+    /// </summary>
+    public ILoggerFactory? LoggerFactory { get; set; } = null;
```
Comment (lemillermicrosoft) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlannerConfig.cs:85]: I added this before when it was a Logger. Probably makes sense to remove this and just use the kernel logger factory. Thoughts appreciated.

Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlannerConfig.cs:85]: Removing for now. Will add if folks think appropriate.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 5]
```diff
@@ -98,95 +99,286 @@ public Plan CreatePlan(string goal)
...
+
+    private Task<string> GetSystemMessage(SKContext context)
+    {
+        return this._promptRenderer.RenderAsync(this._promptTemplate, context);
+    }
+
+    #endregion setup helpers
+
+    #region execution helpers
     public virtual SystemStep ParseResult(string input)
```
Comment (lemillermicrosoft) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:382]: Does this really need to be public? We should rename to make it more clear what it's doing and add some xml comments if so.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 6]
```diff
@@ -98,95 +99,286 @@ public Plan CreatePlan(string goal)
...
-        return context;
+        return chatHistory;
+    }
+
+    private static bool TryGetChatCompletion(IKernel kernel, [NotNullWhen(true)] out IChatCompletion? chatCompletion)
+    {
+        try
+        {
+            // Client used to request answers to chat completion models
+            chatCompletion = kernel.GetService<IChatCompletion>();
```
Comment (andhesky) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:352]: >GetService [](http://example.com/codeflow?start=36&length=10)

Using exception handling in the normal flow is fairly expensive. Can we add a TryGetService to the kernel API?

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 7]
```diff
@@ -98,95 +99,286 @@ public Plan CreatePlan(string goal)
...
+#pragma warning disable CA2016 // Forward the 'CancellationToken' parameter to methods that take one -- the method is obsolete and to be removed.
+        var systemContext = this._kernel.CreateNewContext();
+#pragma warning restore CA2016
+
+        systemContext.Variables.Set("suffix", this.Config.Suffix);
+
+        if (chatCompletion is null)
+        {
+            systemContext.Variables.Set("functionDescriptions", userManual);
+        }
```
Comment (andhesky) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:316]: I think this is already set in GetUserManualAsync(). Why do we need to set it again?

Comment (andhesky) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:316]: I don't understand what this is doing. We overwrite the function descriptions with the userManual. Why?

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 8]
```diff
@@ -98,95 +99,286 @@ public Plan CreatePlan(string goal)
...
+        string systemMessage = await this.GetSystemMessage(systemContext).ConfigureAwait(false);
+
+        chatHistory.AddSystemMessage(systemMessage);
+        if (chatCompletion is not null)
+        {
+            chatHistory.AddUserMessage(userManual);
+        }
```
Comment (andhesky) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:323]: Why the special case? At least needs a comment.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs (hunk 7 above)
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:316]: Will add explanation, cause it does appear confusing -- and may be worth changing.

In short -- I found when using Chat Completion, that putting the available functions as the first user message, instead of injected into the system message, to have better results (the LLM would respect it and hallucinate less leading to fewer iterations to answer asks). The more I learn about function calling support, and the fact it injects into the system message behind the scenes, makes me question my anecdotal experience.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 9]
```diff
@@ -98,95 +99,286 @@ public Plan CreatePlan(string goal)
...
+                {
+                    throw new SKException("No completions returned.");
+                }
 
-                context.Variables.Set("agentScratchPad", scratchPad);
+                return await results[0].GetCompletionAsync(token).ConfigureAwait(false);
+            }
+
+            throw new SKException("No AIService available for getting completions.");
+        };
```
Comment (andhesky) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:147]: Having a wrapper around the text completion as in my PR feels like the cleaner solution to me. This works, but risks having the code accidentally diverge in the future.

If you don't want to implement a full fledged wrapper you could implement it as a private class internal to StepwisePlanner for now and consider making it public in the future.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs (hunk 6 above)
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:352]: Opened #2635 to track and will leave note here for now.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 10]
```diff
@@ -98,95 +99,286 @@ public Plan CreatePlan(string goal)
...
+            var nextStep = await GetNextStepAsync().ConfigureAwait(false);
+
+            // If not Action/Thought/FinalAnswer is found,
+            // add a message to the chat history to guide LLM into returning the next thought.
+            if (string.IsNullOrEmpty(nextStep.Action) &&
+                string.IsNullOrEmpty(nextStep.Thought) &&
+                string.IsNullOrEmpty(nextStep.FinalAnswer))
+            {
+                this._logger?.LogWarning("No response from LLM");
+                chatHistory?.AddUserMessage(Thought);
```
Comment (andhesky) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:200]: >AddUserMessage [](http://example.com/codeflow?start=29&length=14)

Why user message and not assistant?

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs (hunk 5 above)
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:382]: making this protected internal

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs (hunk 7 above)
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:320]: Added comment. Going to run a comp right now. Appreciate folks thoughts on default behavior.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs (hunk 9 above)
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:165]: cc @RogerBarreto if they have any thoughts on creating that kind of wrapper, #2444  for context.

I'm not sure if we want to be prescriptive or not when passing a chathistory to a text completion endpoint. And while we could leverage such an interface to allow users to use StepwisePlanner with their own converters, I don't know if that's the stance we want to take with the kernel. These planners are extensions of the kernel and demonstrations of what can be built using kernel completion services and skill collections. And while we aim for them to be useful for many scenarios, we admit we are not going to be able to build a single planner that is capable of every scenario.

Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:165]: Definitely open to incremental improvements here, too, so opening #2640 to track this.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs (hunk 7 above)
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:330]: Okay, yeah, I'm not seeing that difference in these examples. I'm going to revert this special logic.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 11]
```diff
@@ -94,100 +95,277 @@ public Plan CreatePlan(string goal)
...
-            for (int i = 0; i < this.Config.MaxIterations; i++)
+            var messages = string.Join("\n", chatHistory?.Messages);
+            var tokenCount = messages.Length / 4;
+            return tokenCount;
+        };
+
+        var addThought = true;
+        var GetCompletionAsync = async () =>
+        {
+            if (chatCompletion is not null)
```
Comment (SergeyMenshykh) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:137]: No need to declare `chatCompletion` above and check if it has IChatCompletion type. It can easily be done in one line - `if (aiService is IChatCompletion chatCompletion)`

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 12]
```diff
@@ -94,100 +95,277 @@ public Plan CreatePlan(string goal)
...
+    /// <summary>
+    /// Execute a plan
+    /// </summary>
+    /// <param name="question">The question to answer</param>
+    /// <param name="context">The context to use</param>
+    /// <param name="token">The cancellation token</param>
+    /// <returns>The context with the result</returns>
+    /// <exception cref="SKException">No AIService available for getting completions.</exception>
     [SKFunction, SKName("ExecutePlan"), Description("Execute a plan")]
     public async Task<SKContext> ExecutePlanAsync(
```
Comment (SergeyMenshykh) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:107]: Please split the method into smaller ones for better readability and maintenance.

Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:107]: Yeah, it's a bit unwieldly.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs (hunk 9 above)
Comment (andhesky) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:165]: Makes sense to decouple. I'll take a stab at that wrapper because it turns out I may need it soon.

If callers want their own implementation they can always write their own wrapper and register it as a chat completion and it will be picked up by this planner.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs (hunk 12 above)
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:107]: Okay, putting a stake in this one for now.  I've started to go down a few paths for future improvement:

- Define JSON schema for Action and allow Action[] to be returned
- Extract the engine components into their own class to remove coupling with chat history
- Reconsider whether we show invalid results as assistant messages to encourage learning at the trade off of keeping it aligned to expectations.
- Significant test automation to reduce manual burdens.

**Review by SergeyMenshykh — APPROVED** [review]

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs (hunk 12 above)
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:107]: cc @hario90

File: dotnet/samples/KernelSyntaxExamples/Example51_StepwisePlanner.cs [hunk 13]
```diff
@@ -12,77 +15,213 @@
...
-    private static async Task RunWithQuestion(IKernel kernel, string question)
+    private struct ExecutionResult
     {
+        public string mode;
+        public string? model;
+        public string? question;
+        public string? answer;
+        public string? stepsTaken;
+        public string? iterations;
+        public string? timeTaken;
```
Comment (dmytrostruk) [inline dotnet/samples/KernelSyntaxExamples/Example51_StepwisePlanner.cs:84]: nit: Should it be PascalCase to follow naming convention?

File: dotnet/samples/KernelSyntaxExamples/Example51_StepwisePlanner.cs [hunk 14]
```diff
@@ -12,77 +15,213 @@
...
+        if (useChat)
+        {
+            builder.WithAzureChatCompletionService(
+                model ?? ChatModelOverride ?? TestConfiguration.AzureOpenAI.ChatDeploymentName,
+                TestConfiguration.AzureOpenAI.Endpoint,
+                TestConfiguration.AzureOpenAI.ApiKey,
+                alsoAsTextCompletion: true,
+                setAsDefault: true);
+
+            Console.WriteLine($"Model: {model ?? ChatModelOverride ?? TestConfiguration.AzureOpenAI.ChatDeploymentName} ({ChatMaxTokens ?? (new Microsoft.SemanticKernel.Planning.Stepwise.StepwisePlannerConfig()).MaxTokens})");
```
Comment (dmytrostruk) [inline dotnet/samples/KernelSyntaxExamples/Example51_StepwisePlanner.cs:212]: Improve formatting for better readability. It's important since example code should be easy to follow.

File: dotnet/samples/KernelSyntaxExamples/Example51_StepwisePlanner.cs [hunk 15]
```diff
@@ -12,77 +15,213 @@
...
+                TestConfiguration.AzureOpenAI.Endpoint,
+                TestConfiguration.AzureOpenAI.ApiKey);
 
-        builder.WithAzureChatCompletionService(
-            TestConfiguration.AzureOpenAI.ChatDeploymentName,
-            TestConfiguration.AzureOpenAI.Endpoint,
-            TestConfiguration.AzureOpenAI.ApiKey,
-            alsoAsTextCompletion: true,
-            setAsDefault: true);
+            Console.WriteLine($"Model: {model ?? TextModelOverride ?? TestConfiguration.AzureOpenAI.DeploymentName} ({TextMaxTokens ?? (new Microsoft.SemanticKernel.Planning.Stepwise.StepwisePlannerConfig()).MaxTokens})");
```
Comment (dmytrostruk) [inline dotnet/samples/KernelSyntaxExamples/Example51_StepwisePlanner.cs:222]: Improve formatting for better readability. It's important since example code should be easy to follow.

File: dotnet/src/Extensions/Extensions.StepwisePlanner.UnitTests/Planning/StepwisePlanner/ParseResultTests.cs [hunk 16]
```diff
@@ -35,11 +38,36 @@ public void WhenInputIsFinalAnswerReturnsFinalAnswer(string input, string expect
...
+    [InlineData("The web search result is a snippet from a Wikipedia article that says something.\n\n[ACTION] {\n  \"action\": \"WebSearch.Search\",\n \"action_variables\": {\"input\": \"another search\", \"count\": \"1\"}\n}", "The web search result is a snippet from a Wikipedia article that says something.", "WebSearch.Search", "input",
+        "another search", "count", "1")]
+    [InlineData("[ACTION] {\"action\": \"time.Year\", \"action_variables\": {\"input\": \"\"}}", null, "time.Year", "input", "")]
+    [InlineData(@"[ACTION]{
+  ""action"": ""RepositorySkill.PushChangesToBranch"",
+  ""action_variables"": {
+    ""branchName"": ""myBranchName"",
+    ""comment"": ""{MyComment""
+  }
+}
```
Comment (dmytrostruk) [inline dotnet/src/Extensions/Extensions.StepwisePlanner.UnitTests/Planning/StepwisePlanner/ParseResultTests.cs:52]: Idea for follow-up PR:
It will be hard to maintain these test cases. Instead of `InlineData` we can use `ClassData`, which will allow us to use dedicated class for test cases, so we can keep them in separate place or read it from .txt file.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 17]
```diff
@@ -94,255 +95,454 @@ public Plan CreatePlan(string goal)
...
+            throw new SKException("ChatHistory is null.");
+        }
+
+        var startingMessageCount = chatHistory.Messages.Count;
+
         var stepsTaken = new List<SystemStep>();
-        if (!string.IsNullOrEmpty(question))
+        SystemStep? lastStep = null;
+
+        var GetNextStepAsync = async () =>
```
Comment (dmytrostruk) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:136]: I'm wondering if it would be possible to keep all these delegates as separate methods to simplify root `ExecutePlanAsync` method.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 18]
```diff
@@ -94,255 +95,454 @@ public Plan CreatePlan(string goal)
...
+                result.Thought = thoughtMatch.Value.Trim();
             }
+        }
+        else if (!input.Contains(Action))
+        {
+            result.Thought = input;
+        }
+        else
+        {
+            // result.Observation = "Could not parse response. Unable to find thought or action. Please try again.";
```
Comment (dmytrostruk) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:489]: nit: Should we remove this? 
```suggestion
```

**Review by dmytrostruk — APPROVED** [review]
Left a couple of small comments, overall LGTM. 
Feel free to resolve my comments and address them in follow-up PRs.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs (hunk 17 above)
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:136]: Agreed, started to go down this path. I think the right thing to do will be creating a Parser and Engine to simplify responsibilities.

File: dotnet/samples/KernelSyntaxExamples/Example51_StepwisePlanner.cs (hunk 13 above)
Comment (lemillermicrosoft) [reply dotnet/samples/KernelSyntaxExamples/Example51_StepwisePlanner.cs:84]: Thanks Dmytro, filing a follow up task for these pieces.

**Review by dmytrostruk — APPROVED** [review]
